// .ignored/@ngxs/store/esm2022/src/module.mjs
import { NgModule } from '@angular/core';
import { NgxsRootModule } from './modules/ngxs-root.module';
import { NgxsFeatureModule } from './modules/ngxs-feature.module';
import { getRootProviders } from './standalone-features/root-providers';
import { getFeatureProviders } from './standalone-features/feature-providers';
import * as i0 from "@angular/core";
export class NgxsModule {
    static forRoot(states = [], options = {}) {
        return {
            ngModule: NgxsRootModule,
            providers: getRootProviders(states, options)
        };
    }
    static forFeature(states = []) {
        return {
            ngModule: NgxsFeatureModule,
            providers: getFeatureProviders(states)
        };
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: NgxsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.6", ngImport: i0, type: NgxsModule }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: NgxsModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: NgxsModule, decorators: [{
            type: NgModule
        }] });
