// @ngxs/storage-plugin@18.1.1/esm2022/src/internals.mjs
import { isPlatformServer } from '@angular/common';
import { ɵDEFAULT_STATE_KEY } from '@ngxs/storage-plugin/internals';
export function storageOptionsFactory(options) {
    return {
        storage: 0 /* StorageOption.LocalStorage */,
        serialize: JSON.stringify,
        deserialize: JSON.parse,
        beforeSerialize: obj => obj,
        afterDeserialize: obj => obj,
        ...options,
        keys: options.keys === '*' ? [ɵDEFAULT_STATE_KEY] : options.keys
    };
}
export function engineFactory(options, platformId) {
    if (isPlatformServer(platformId)) {
        return null;
    }
    if (options.storage === 0 /* StorageOption.LocalStorage */) {
        return localStorage;
    }
    else if (options.storage === 1 /* StorageOption.SessionStorage */) {
        return sessionStorage;
    }
    return null;
}
export function getStorageKey(key, options) {
    // Prepends the `namespace` option to any key if it's been provided by a user.
    // So `@@STATE` becomes `my-app:@@STATE`.
    return options?.namespace ? `${options.namespace}:${key}` : key;
}
