// @ngxs/storage-plugin@18.1.1/esm2022/src/storage.module.mjs
import { NgModule, PLATFORM_ID, makeEnvironmentProviders } from '@angular/core';
import { withNgxsPlugin } from '@ngxs/store';
import { NGXS_PLUGINS } from '@ngxs/store/plugins';
import { ɵUSER_OPTIONS, STORAGE_ENGINE, ɵNGXS_STORAGE_PLUGIN_OPTIONS } from '@ngxs/storage-plugin/internals';
import { NgxsStoragePlugin } from './storage.plugin';
import { engineFactory, storageOptionsFactory } from './internals';
import * as i0 from "@angular/core";
export class NgxsStoragePluginModule {
    static forRoot(options) {
        return {
            ngModule: NgxsStoragePluginModule,
            providers: [
                {
                    provide: NGXS_PLUGINS,
                    useClass: NgxsStoragePlugin,
                    multi: true
                },
                {
                    provide: ɵUSER_OPTIONS,
                    useValue: options
                },
                {
                    provide: ɵNGXS_STORAGE_PLUGIN_OPTIONS,
                    useFactory: storageOptionsFactory,
                    deps: [ɵUSER_OPTIONS]
                },
                {
                    provide: STORAGE_ENGINE,
                    useFactory: engineFactory,
                    deps: [ɵNGXS_STORAGE_PLUGIN_OPTIONS, PLATFORM_ID]
                }
            ]
        };
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: NgxsStoragePluginModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.0.6", ngImport: i0, type: NgxsStoragePluginModule }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: NgxsStoragePluginModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: NgxsStoragePluginModule, decorators: [{
            type: NgModule
        }] });
export function withNgxsStoragePlugin(options) {
    return makeEnvironmentProviders([
        withNgxsPlugin(NgxsStoragePlugin),
        {
            provide: ɵUSER_OPTIONS,
            useValue: options
        },
        {
            provide: ɵNGXS_STORAGE_PLUGIN_OPTIONS,
            useFactory: storageOptionsFactory,
            deps: [ɵUSER_OPTIONS]
        },
        {
            provide: STORAGE_ENGINE,
            useFactory: engineFactory,
            deps: [ɵNGXS_STORAGE_PLUGIN_OPTIONS, PLATFORM_ID]
        }
    ]);
}
