// @ngxs/store@18.1.1/esm2022/internals/state-stream.mjs
import { Injectable, untracked } from '@angular/core';
import { toSignal } from '@angular/core/rxjs-interop';
import { ɵwrapObserverCalls } from './custom-rxjs-operators';
import { ɵOrderedBehaviorSubject } from './custom-rxjs-subjects';
import * as i0 from "@angular/core";
/**
 * BehaviorSubject of the entire state.
 * @ignore
 */
export class ɵStateStream extends ɵOrderedBehaviorSubject {
    constructor() {
        super({});
        this.state = toSignal(this.pipe(ɵwrapObserverCalls(untracked)), {
            manualCleanup: true,
            requireSync: true
        });
    }
    ngOnDestroy() {
        // The StateStream should never emit values once the root view is removed,
        // such as when the `NgModuleRef.destroy()` method is called. This is crucial
        // for preventing memory leaks in server-side rendered apps, where a new StateStream
        // is created for each HTTP request. If users forget to unsubscribe from `store.select`
        // or `store.subscribe`, it can result in significant memory leaks in SSR apps.
        this.complete();
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: ɵStateStream, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    /** @nocollapse */ static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: ɵStateStream, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.6", ngImport: i0, type: ɵStateStream, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: () => [] });
